api/auth.proto

```diff
@@ -0,0 +1,48 @@
+syntax = "proto3";
+
+package envoy.api.v2;
+
+message AuthAction {
+  // Should we do white-list or black-list style access control.
+  enum ActionType {
+    // Request matches all rules are allowed, otherwise denied.
+    ALLOW = 0;
+    // Request matches all rules or missing required auth fields are denied,
+    // otherwise allowed.
+    DENY = 1;
+  }
+
+  ActionType action_type = 1;
+
+  // Element type of AndRule/OrRule, it chooses among different type of rule.
+  message Rule {
+    oneof rule_oneof {
+      AndRule and_rule = 1;
+      OrRule or_rule = 2;
+      X509Rule x509_rule = 3;
+    }
+  }
+
+  // Logic AND that requires all rules match.
+  message AndRule {
+    repeated Rule rules = 1;
+  }
+
+  // Logic OR that requires at least one rule matches.
+  message OrRule {
+    repeated Rule rules = 1;
+  }
+
+  // Check peer identity using X.509 certificate.
+  message X509Rule {
+    // SHA256 hash of peer certificate.
+    string sha256_hash = 1;
+    // Subject of peer certificate.
+    string subject = 2;
+    // Subject alt name of peer certificate.
+    string subject_alt_name = 3;
+  }
```

Comment:
Again (and sorry if this was already answered, but I cannot find the answer) - why we don't use `CertificateValidationContext` here?

Reply:
I think the idea is that `CertificateValidationContext` is a large superset of what we want here. I had the comment:

"As a contrived example, in the same way in which I would want to verify a particular route as having a particular SAN, might I not always want to ensure I reached the route via a listener that performed validation via a specific CA? It sounds contrived to me and not very realistic, but other folks probably have opinions."

So, yeah, maybe we can get some final thoughts on where we want to go here and reach consensus.

Reply:
I just looked and I agree with @PiotrSikora. We should probably use `CertificateValidationContext` here. @jwfang when you make that change do you mind also merging master so that tests run?